core/server/data/migrations/versions/5.0/2022-04-21-15-44-backfill-mrr-events-for-canceled-subscriptions.js

```diff
@@ -0,0 +1,77 @@
+const ObjectID = require('bson-objectid').default;
+const logging = require('@tryghost/logging');
+
+const {createTransactionalMigration} = require('../../utils');
+
+module.exports = createTransactionalMigration(
+    async function up(knex) {
+        const canceledSubscriptions = await knex('members_stripe_customers_subscriptions')
+            .select(
+                'members_stripe_customers_subscriptions.id',
+                'members_stripe_customers_subscriptions.updated_at',
+                'members_stripe_customers_subscriptions.mrr',
+                'members_stripe_customers_subscriptions.stripe_price_id',
+                'members_stripe_customers_subscriptions.plan_currency',
+                'members_stripe_customers.member_id'
+            )
+            .join('members_stripe_customers', 'members_stripe_customers_subscriptions.customer_id', '=', 'members_stripe_customers.customer_id')
+            .join('members', 'members_stripe_customers.member_id', '=', 'members.id')
+            .where('cancel_at_period_end', '=', true)
+            .where('mrr', '!=', 0);
+
+        if (canceledSubscriptions.length === 0) {
+            logging.info('No canceled subscriptions found, skipping migration.');
+            return;
+        } else {
+            logging.info(`Found ${canceledSubscriptions.length} canceled subscriptions, updating MRR events`);
+        }
+
+        const canceledEvents = await knex('members_paid_subscription_events')
+            .select(
+                'id',
+                'subscription_id'
+            )
+            .where('type', '=', 'canceled')
+            .whereIn('subscription_id', canceledSubscriptions.map(x => x.id))
```

Comment:
Is this safe given that we don't know how many entries we might have?

Reply:
As long as we haven't got more than 32k we should be good!